documentation/docs/protocol_specification/components/congestion_control.md

```diff
@@ -1,87 +1,164 @@
-*   _Message drop_. In some circumstances, due to network congestion or to ongoing attacks, some messages shall be dropped to guarantee bounded delays and isolate attacker's messages. Specifically, a node shall drop messages in two situations:
-    - since buffers are of a limited size, if the total number of bytes in all queues exceeds a certain threshold, new incoming messages are dropped;
-    - to guarantee the security of the network, if a certain queue exceeds a given threshold, new incoming packets from that specific node ID will be dropped.
-
-The dequeue mechanism includes the following components:
+* __Classification__: The mechanism identifies the queue where the message belongs according to the node ID of
+  the message issuer.
+* __Message enqueuing__: The message is actually enqueued, the queue is sorted by message timestamps in increasing order
+  and counters are updated (e.g., counters for the total number of bytes in the queue).
 
-*   _Queue selection_. A queue is selected according to round robin scheduling algorithm. In particular, we use a modified version of the deficit round robin (DRR) algorithm.
-*   _Message dequeuing_. The first message of the queue is dequeued, and list of active nodes is updated.
-*   _Scheduler management_. Scheduler counters and pointers are updated.
+The dequeuing mechanism includes the following components:
 
-### Scheduler
-
-The most critical task is the scheduling algorithm which must guarantee that, for an honest node `node`, the following requirements will be met:
-* `node`'s messages will not accumulate indefinitely at any node (i.e., starvation is avoided), so the _consistency_ requirement will be ensured.
-* `node`'s fair share (according to its access Mana) of the network resources are allocated to it, guaranteeing the _fairness_ requirement.
-* Malicious nodes sending above their allowed rate will not interrupt `node`'s throughput, fulfilling the _security_ requirement.
+* __Queue selection__: A queue is selected according to a round-robin scheduling algorithm. In particular, the 
+mechanism uses a modified version of the deficit round-robin (DRR) algorithm.
+* __Message dequeuing__. The first (oldest) message of the queue, that satisfies certain conditions is dequeued. A 
+  message must satisfy the following conditions:
+    * The message has a ready flag assigned. Ready flag is assigned to a message when all of its parents are eligible (parents have been scheduled or confirmed).
+    * The message timestamp is not in the future.
+* __Message skipping__. Once a message in the outbox becomes confirmed, that is we receive another message approving it, it gets removed from the outbox buffer. Since the message already has approvers and is supposed to be replicated on enough nodes in the network, it is not gossiped or added to the tip pool, hence "skipped".
```

Comment:
```suggestion
* __Message skipping__. Once a message in the outbox is confirmed by another message approving it, it will get removed from the outbox buffer. Since the message already has approvers and is supposed to be replicated on enough nodes in the network, it is not gossiped or added to the tip pool, hence "skipped".
```